---
title: "CENTRE-vignette"
author: 
- name: Sara Lopez Ruiz de Vargas
  affiliation: Mantainer
  email: lopez_s@molgen.mpg.de
- name: Trisevgeni Rapakoulia
  affiliation: Creator
  email: rapakoul@molgen.mpg.de
package: CENTRE
output: 
  rmarkdown::html_document
abstract: 
vignette: |
  %\VignetteIndexEntry{Centre-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      comment = "#>")
```

# Overview
The *CENTRE* pipeline follows the workflow below. Two use cases :

1. User has only genes: `createPairs()` -> `computeGenericFeatures()` ->
`computeCellTypeFeatures()` -> `centreClassification()`

2. User has gene enhancer pairs : `computeGenericFeatures()` ->
 `computeCellTypeFeatures()` -> `centreClassification()`
 
# Getting started
In order to run *CENTRE* you will need the following :

- Cell-type specific histone ChIP-seq in BAM format for H3K27ac, H3K4me3 and 
H3K4me1. Additionally, a Control ChIP-seq experiment to match the HM ChIP-seq is
but CENTRE can also run without it.
- Cell -type specific RNA-seq TPM values for all genes. This dataframe will have
three columns one with the ENSEMBL ID's, transcript ID's and one with the TPM 
values for all genes.
- A dataframe with either the GENCODE ID's for the genes of interest or enhancer
(cCREs-ELS) and target (GENCODE ID's) pairs of interest.

The first time you run CENTRE you will need to download the PrecomputedData.db 
and Annotation.db. To do this run the following: 

```{r}
CENTRE::downloadPrecomputedData(method = "curl")
# Make sure whatever method you use to download is available on your system
```
Or download the data from http://owww.molgen.mpg.de/~CENTRE_data/PrecomputedData.db
and http://owww.molgen.mpg.de/~CENTRE_data/Annotation.db add it to the /inst/extdata folder.

# Run the *CENTRE* pipeline

## Example Data
We provide an in-package example on the cell line HeLa-S3. The Histone Mark
and RNA-seq data are from ENCODE and correspond to the following experiments:


| Experiment type| ENCODE experiment accession | File accession numbers|
|----------------|-----------------------------|-----------------------|
| H3K4me1        |ENCSR000APW| ENCFF712AAP, ENCFF826OLG|
| H3K4me3        | ENCSR340WQU|ENCFF650IXI, ENCFF760VTC |
| H3K27ac        |ENCSR000AOC | ENCFF609ZAE, ENCFF711QAI |
| ChIP-seq Control| ENCSR000AOB|ENCFF017QCL, ENCFF842IEZ|
| RNA-seq        | ENCSR000CPP |ENCFF297BJF, ENCFF623UDU


For the Histone ChIP-seq replicates were merged using bamtools merge, separated
by chromosome using bamtools split and indexed using bamtools index.
For RNA-seq data we take the mean of the TPM values over 
the replicates. 

## Step 1: Create gene enhancer pairs
If you are in use case one you will need to use `CENTRE::createPairs()`. This
function fins all enhancers in 500kb from the transcription start site of the
input genes and creates all possible enhancer gene pairs.

It takes as input a dataframe `genes` with one column of ENSEMBL gene ID's.
The output is a dataframe of two columns one with the gene ID's (without version
identifier) and the corresponding ENCODE cCREs enhancer ID's.
```{r}
genes <- as.data.frame(c("ENSG00000130203.10",
                         "ENSG00000171119.3"))
colnames(genes) <- c("gene_id") #It is important to name the column gene_id
pairs <- CENTRE::createPairs(genes)
```

## Step 2: Compute Generic Features
The function `computeGenericFeatures(pairs)` computes the features that are not
cell type specific.

Takes as input a dataframe `pairs` with a format like the one in the last table.
The function returns a dataframe with the following features as columns:

- `distance`: the distance between the middle point of the enhancer and the
transcription start site of the gene in each pair
- `cor_CRUP`: the CRUP correlation for each pair. This is one of the precomputed
datasets.
- `combined_tests`: the combined p-values of all of the precomputed Wilcoxon
tests.

```{r}
colnames(pairs) <- c("gene_id", "enhancer_id")
generic_features <- CENTRE::computeGenericFeatures(pairs)
```

## Step 3: Compute Cell Type Features
The function `computeCellTypeFeatures(metaData, replicate, input.free, sequencing, tpm, featuresGeneric)` computes the features that are cell type specific.
Takes as input the following

- `metaData`: Dataframe with the path to the cell type specific ChIP-seq
experiments.
- `replicate`: The number of replicates of the ChIP-seq experiments
that need to be normalized.
- `input.free`: Boolean value indicating whether a Control/Input ChIP-seq
experiment is provided to go with the Histone Modification ChIP-seq experiments.
If the parameter is set to FALSE the normalization of ChIP-seq experiments
will be run in input.free mode.
- `cores`: Integer indicating how many cores CRUP should be run with.
- `sequencing`: String "single" or "paired" indicating the type of sequencing of
the histone ChIP-seq experiments.
- `tpm`: RNA-seq gene quantification data in R dataframe format. The R 
dataframe has to have 3 columns, one for the `gene_id`, 
one for the `transcript_ids` and one for the TPM value, for all genes and not
just the genes of interest.
- `features_generic`: The dataframe that was produced in `computeGenericFeatures()`

Example of the tpm dataframe: 

|gene_id | transcript_id.s.| TPM|
|--------|-----------------|----|
|   10904 |           10904|   0|
|   12954|            12954|   0|
|   12956|            12956|   0|
|   ... |            ...|   0|

Returns the cell type specific features : 

- `EP_prob_enh`: CRUP-EP(Enhancer Probability) for Enhancers
- `EP_prob_gene`: CRUP-EP(Enhancer Probability) for Promoters
- `reg_dist_enh`: Regulatory distance computed on enhancer probabilities
- `norm_reg_dist_enh`: Normalized regulatory distance computed on enhancer probabilities
- `PP_prob_enh`: CRUP-PP(Promoter Probability) for Enhancers
- `PP_prob_gene`: CRUP-PP(Promoter Probability) for Promoters
- `reg_dist_enh`: Regulatory distance computed on promoter probabilities
- `norm_reg_dist_enh`: Normalized regulatory distance computed on promoter probabilities
- `RNA_seq`: RNA-seq TPM values for the gene in each pair of the cell type of interest.

First prepare the metadata and tpm dataframes and then run the function.
```{r}
files <- c(system.file("extdata/example","HeLa_H3K4me1.REF_chr19.bam", package = "CENTRE"),
          system.file("extdata/example","HeLa_H3K4me3.REF_chr19.bam", package = "CENTRE"),
          system.file("extdata/example","HeLa_H3K27ac.REF_chr19.bam", package = "CENTRE"))
          
# Control ChIP-seq experiment to go with the rest of ChIP-seqs
inputs <- system.file("extdata/example", "HeLa_input.REF_chr19.bam", package = "CENTRE")

metaData <- data.frame(HM = c("H3K4me1", "H3K4me3", "H3K27ac"),
                       condition = c(1, 1, 1), replicate = c(1, 1, 1),
                       bamFile = files, inputFile = rep(inputs, 3))


tpmfile <- read.table(system.file("extdata/example", "HeLa-S3.tsv", package = "CENTRE"),
                      sep = "", stringsAsFactors = F, header = T)

celltype_features <- CENTRE::computeCellTypeFeatures(metaData,
                                    replicate = 1,
                                    input.free = FALSE,
                                    cores = 1,
                                    sequencing = "single",
                                    tpmfile = tpmfile,
                                    featuresGeneric = generic_features)

```

### Step 4: Classify Gene Enhancer Pairs

With the function `centrePrediction(features_generic, features_celltype, model)`
the gene enhancer targets are classified as active or inactive.
The function takes as input the generic features, the cell type specific features
and the model (CENTRE model by default).

Returns a dataframe with the `pairs` ID `enhancer_id` and `gene_id` concatenated
by a `_` of the corresponding pair and the label and probability for each of the pairs.

```{r}
# Finally compute the predictions
predictions <- centrePrediction(celltype_features,
                                  generic_features)
```

